onnxruntime/core/providers/cuda/cuda_execution_provider.cc

```diff
@@ -2431,35 +2431,46 @@ CUDAExecutionProvider::GetCapability(const onnxruntime::GraphViewer& graph,
   return result;
 }
 
-void CUDAExecutionProvider::RegisterAllocator(std::shared_ptr<AllocatorManager> allocator_manager) {
+void CUDAExecutionProvider::RegisterAllocator(AllocatorManager& allocator_manager) {
+  OrtDevice cuda_device{OrtDevice::GPU, OrtDevice::MemType::DEFAULT, info_.device_id};
+  OrtDevice pinned_device{OrtDevice::CPU, OrtDevice::MemType::CUDA_PINNED, DEFAULT_CPU_ALLOCATOR_DEVICE_ID};
+  OrtDevice cpu_device{OrtDevice::CPU, OrtDevice::MemType::DEFAULT, DEFAULT_CPU_ALLOCATOR_DEVICE_ID};
+
   // Try to get a CUDA allocator from allocator manager first
   // Used to allocate CUDA device memory
-  auto cuda_alloc = allocator_manager->GetAllocator(info_.device_id, OrtMemTypeDefault);
+  auto cuda_alloc = allocator_manager.GetAllocator(OrtMemTypeDefault, cuda_device);
   if (nullptr == cuda_alloc) {
     cuda_alloc = CreateCudaAllocator(info_.device_id, info_.gpu_mem_limit, info_.arena_extend_strategy,
                                      info_.external_allocator_info, info_.default_memory_arena_cfg);
-    allocator_manager->InsertAllocator(cuda_alloc);
+    allocator_manager.InsertAllocator(cuda_alloc);
   }
+
   TryInsertAllocator(std::move(cuda_alloc));
```

Comment:
I don't understand why we need TryInsertAllocator in these EPs. This should be the only place we're inserting an allocator into the per-EP list, and RegisterAllocator is only called once per EP in the inference session. Is there a scenario where we need this? 

Only one that comes to mind is if someone registers two instances of the same EP with the same device_id (as the device info is part of the key).

Reply:
The CIs I tried with InsertAllocator passed, so using that unless there's a need to support the 'try' approach.